### ISSUE TITLE
Viewing a revision results in 403 error

### ISSUE THREAD

**Issue by andybroomfield** [localgovdrupal/localgov_alert_banner#171]
When looking at revisions, its not actully possible to view a revision.
This is even with user 1 account.

Steps to reproduce:
1. on an alert banner entity, go to revisions tab
2. click a revision that is not the latest
3. witness 403 access denied error

What should happen:
user is able to view revision.

This was found while checking #169 but I think the error is wider.
Localgov Workflows was installed at the time.

**Comment by andybroomfield** [comment]
Might be releated to #151?

**Comment by andybroomfield** [comment]
Message | Path: /admin/content/alert-banner/8/revisions/24/view. Drupal\Core\Http\Exception\CacheableAccessDeniedHttpException: in Drupal\Core\Routing\AccessAwareRouter->checkAccess() (line 117 of /app/web/core/lib/Drupal/Core/Routing/AccessAwareRouter.php).
-- | --

Debugging this gives 
> stdClass Object ( [__CLASS__] => Drupal\Core\Access\AccessResultNeutral [reason:protected] => [cacheContexts:protected] => Array ( ) [cacheTags:protected] => Array ( ) [cacheMaxAge:protected] => -1 )

**Comment by andybroomfield** [comment]
Maybe releated Drupal core issue? https://www.drupal.org/project/drupal/issues/2991698

**Comment by andybroomfield** [comment]
From discovery, this is the code that is ultimatly responsible for checking access. It is in class Drupal\Core\Entity\EntityAccessCheck.

What is happening is that it is checking for an entity, but the router for the revision is providing just the entity ID so it return access neutral, which in turn causes the access denined.

Is this a Localgov issue or a core issue? I wasn't able to access this using the media entities so couldn't check.
Checking with some other custom entities on BHCC does allow the revision to be viewed.

````php
  /**
   * Checks access to the entity operation on the given route.
   *
   * The route's '_entity_access' requirement must follow the pattern
   * 'entity_stub_name.operation', where available operations are:
   * 'view', 'update', 'create', and 'delete'.
   *
   * For example, this route configuration invokes a permissions check for
   * 'update' access to entities of type 'node':
   * @code
   * pattern: '/foo/{node}/bar'
   * requirements:
   *   _entity_access: 'node.update'
   * @endcode
   * And this will check 'delete' access to a dynamic entity type:
   * @code
   * example.route:
   *   path: foo/{entity_type}/{example}
   *   requirements:
   *     _entity_access: example.delete
   *   options:
   *     parameters:
   *       example:
   *         type: entity:{entity_type}
   * @endcode
   * The route match parameter corresponding to the stub name is checked to
   * see if it is entity-like i.e. implements EntityInterface.
   *
   * @see \Drupal\Core\ParamConverter\EntityConverter
   *
   * @param \Symfony\Component\Routing\Route $route
   *   The route to check against.
   * @param \Drupal\Core\Routing\RouteMatchInterface $route_match
   *   The parametrized route
   * @param \Drupal\Core\Session\AccountInterface $account
   *   The currently logged in account.
   *
   * @return \Drupal\Core\Access\AccessResultInterface
   *   The access result.
   */
  public function access(Route $route, RouteMatchInterface $route_match, AccountInterface $account) {
    // Split the entity type and the operation.
    $requirement = $route->getRequirement('_entity_access');
    list($entity_type, $operation) = explode('.', $requirement);
    // If $entity_type parameter is a valid entity, call its own access check.
    $parameters = $route_match->getParameters();
    if ($parameters->has($entity_type)) {
      $entity = $parameters->get($entity_type);
      if ($entity instanceof EntityInterface) {
        return $entity->access($operation, $account, TRUE);
      }
    }
    // No opinion, so other access checks should decide if access should be
    // allowed or not.
    return AccessResult::neutral();
  }
````

**Comment by stephen-cox** [comment]
@andybroomfield @ekes @finnlewis  I have a partial fix for this issue here: #172 

This core issue helped point the way: https://www.drupal.org/project/drupal/issues/2730631

By adding a routing YAML file defining the route options entities are upcast from the entity and revision IDs to the entity object.

That said the fix in the PR works for viewing revisions but not for reverting or deleting revisions. These will need more investigation and then tests added. I'm unlikely to have time to look at this tomorrow and then off for a bit, so happy for someone else to pick it up if they have time.
